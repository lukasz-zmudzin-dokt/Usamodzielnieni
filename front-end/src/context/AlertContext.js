import React, { useState, createContext } from "react";

export const AlertContext = createContext({
  open: false,
  message: "",
  setOpen: () => {},
});

export const AlertProvider = (props) => {
  const [open, setOpen] = useState(false);
  const [message, setMessage] = useState("");
  const [variant, setVariant] = useState("danger") // success
  const data = {
    open,
    message,
    variant,
    changeVisibility: (newOpen) => setOpen(newOpen),
    changeMessage: (newMessage,newVariant = variant) => {
      setOpen(false);
      setMessage(newMessage);
      setVariant(newVariant);
    },
    showAlert:  (newMessage,newVariant) => {
      data.changeMessage(newMessage,newVariant);
      data.changeVisibility(true);
    }
  };
  return <AlertContext.Provider value={data} {...props} />;
};
