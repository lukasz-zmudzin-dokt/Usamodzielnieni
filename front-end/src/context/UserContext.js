import React, { useState } from "react";
import Cookies from "universal-cookie";

export const UserContext = React.createContext({
  type: undefined, // Employer,Standard,Staff
  token: undefined,
  login: () => {},
  logout: () => {}
});
const cookies = new Cookies();

export const UserProvider = props => {
  const [type, setType] = useState(cookies.get("type"));
  const [token, setToken] = useState(cookies.get("token"));
  const user = {
    token,
    type,
    login: (newToken, newType) => {
      cookies.set("token", newToken, { path: "/" });
      cookies.set("type", newType, { path: "/" });
      setToken(newToken);
      setType(newType);
    },
    logout: () => {
      cookies.remove("token", { path: "/" });
      cookies.remove("type", { path: "/" });
      setToken(undefined);
      setType(undefined);
    }
  };
  return <UserContext.Provider value={user} {...props} />;
};
