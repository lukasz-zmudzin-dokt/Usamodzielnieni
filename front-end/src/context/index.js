export { UserContext, UserProvider } from "context/UserContext";
export { AlertContext, AlertProvider } from "context/AlertContext";
export { ChatContext, ChatProvider } from "context/ChatContext";
export {
  NotificationsContext,
  NotificationsProvider,
} from "context/NotificationsContext";
