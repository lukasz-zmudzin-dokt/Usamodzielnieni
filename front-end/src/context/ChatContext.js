import React, {
  createContext,
  useEffect,
  useState,
  useContext,
  useCallback,
} from "react";
import proxy from "config/api";
import { UserContext } from "context";

const getChats = async (token, filters) => {
  let url = `${proxy.chat}?page=${filters.page}&page_size=${filters.pageSize}`;
  const headers = {
    Authorization: "Token " + token,
    "Content-Type": "application/json",
  };

  const response = await fetch(url, { method: "GET", headers });

  if (response.status === 200) {
    return response.json().then((chats) => chats);
  } else {
    throw response.status;
  }
};

export const ChatContext = createContext(null);

export const ChatProvider = (props) => {
  const user = useContext(UserContext);
  const [chats, setChats] = useState([]);
  const [isChatsLoading, setIsChatsLoading] = useState(false);
  const [error, setError] = useState(false);
  const [count, setCount] = useState(0);
  const [filters] = useState({
    page: 1,
    pageSize: 10,
  });
  const [activePage, setActivePage] = useState(1);

  useEffect(() => {
    if (user.token) {
      const loadChats = async (token) => {
        setActivePage(1);
        setError(false);
        setIsChatsLoading(true);
        let loadedChats;
        try {
          loadedChats = await getChats(token, filters);
        } catch (e) {
          console.log(e);
          loadedChats = { count: 0, results: [] };
          setError(true);
        }
        setChats(loadedChats.results);
        setCount(loadedChats.count);
        setIsChatsLoading(false);
      };
      loadChats(user.token);
    }
  }, [filters, user.token]);

  const loadMoreMessages = async () => {
    let res;
    try {
      res = await getChats(user.token, { ...filters, page: activePage + 1 });
      setActivePage(activePage + 1);
    } catch (e) {
      setError(true);
      return;
    }

    setChats([...chats, ...res.results]);
  };

  const loadMessages = async (token) => {
    setActivePage(1);
    let res;
    try {
      res = await getChats(token, filters);
    } catch (e) {
      setError(true);
      return;
    }

    setChats(res.results);
  };

  const data = {
    chats,
    error,
    count,
    loadMoreMessages,
    isChatsLoading,
    loadMessages,
  };

  return <ChatContext.Provider value={data} {...props} />;
};
