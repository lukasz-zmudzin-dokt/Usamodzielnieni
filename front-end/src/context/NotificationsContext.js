import React, {
  useState,
  useContext,
  useEffect,
  useRef,
  createContext,
} from "react";
import { UserContext } from "context";
import proxy from "config/api";
import { paths } from "constants/paths";
import { userTypes } from "constants/userTypes";
import { compile } from "path-to-regexp";

const getNotifications = async (token, type, next) => {
  let url = proxy.notification + "list/all" + (next ? `?page=${next}` : "");
  const headers = {
    Authorization: "Token " + token,
    "Content-Type": "application/json",
  };

  const response = await fetch(url, { method: "GET", headers });

  if (response.status !== 200) {
    throw response.status;
  }

  return response
    .json()
    .then((notifications) => mapNotifications(notifications, type));
};

const getNotificationsCount = async (token) => {
  let url = proxy.notification + "count/unread";
  const headers = {
    Authorization: "Token " + token,
    "Content-Type": "application/json",
  };

  const response = await fetch(url, { method: "GET", headers });

  if (response.status !== 200) {
    throw response.status;
  }

  return (await response.json()).unread_count;
};

const markAsRead = async (token) => {
  let url = proxy.notification + "list/all/mark-as-read/";
  const headers = {
    Authorization: "Token " + token,
    "Content-Type": "application/json",
  };

  const response = await fetch(url, { method: "POST", headers });

  if (response.status !== 200) {
    throw response.status;
  }

  return;
};

const deleteNotification = async (token, id) => {
  let url = proxy.notification + `delete/${id}/`;
  const headers = {
    Authorization: "Token " + token,
    "Content-Type": "application/json",
  };

  const response = await fetch(url, { method: "DELETE", headers });

  if (response.status !== 200) {
    throw response.status;
  }

  return;
};

const deleteNotifications = async (token) => {
  let url = proxy.notification + "list/all/delete/";
  const headers = {
    Authorization: "Token " + token,
    "Content-Type": "application/json",
  };

  const response = await fetch(url, { method: "DELETE", headers });

  if (response.status !== 200) {
    throw response.status;
  }

  return;
};

const mapNotifications = (notifications, type) => ({
  notifications: notifications.results.map((notification) =>
    mapNotification(notification, type)
  ),
  next: notifications.next?.replace(/.*page=/, "") || undefined,
});

const mapNotification = (notification, type) => ({
  id: notification.slug,
  path: getPath(notification.app, notification.object_id, type),
  title: notification.text,
  time: new Date(notification.timestamp),
  unread: notification.unread,
});

const getPath = (appName, id, type) => {
  const isStandard = type === userTypes.STANDARD;
  if (appName.match(/^cv/)) {
    return isStandard
      ? compile(paths.MY_CVS)({})
      : compile(paths.CV_APPROVAL)({});
  } else if (appName.match(/^jobs/)) {
    return isStandard
      ? compile(paths.JOB_OFFERS)({})
      : compile(paths.MY_OFFERS)({});
  } else if (appName.match(/^account/)) {
    return compile(paths.USER)({});
  } else if (appName.match(/^blog/)) {
    return compile(paths.BLOG_POST)({ id });
  } else {
    return compile(paths.DASHBOARD)({});
  }
};

export const NotificationsContext = createContext(null);

export const NotificationsProvider = (props) => {
  const [notifications, setNotifications] = useState(null);
  const [count, setCount] = useState(null);
  const [next, setNext] = useState();
  const [error, setError] = useState(false);
  const socket = useRef();

  const user = useContext(UserContext);

  useEffect(() => {
    const loadNotifications = async (token, type) => {
      setError(false);
      let values;
      try {
        values = await Promise.all([
          getNotifications(token, type),
          getNotificationsCount(token),
        ]);
      } catch (e) {
        console.log(e);
        setError(true);
        return;
      }
      const [
        { notifications: loadedNotifications, next: loadedNext },
        loadedCount,
      ] = values;
      setCount(loadedCount);
      setNext(loadedNext);
      setNotifications(loadedNotifications);
    };
    if (user.token && user.type) {
      loadNotifications(user.token, user.type);
    }
  }, [user.token, user.type]);

  useEffect(() => {
    if (socket.current) {
      socket.current.close();
      socket.current = null;
    }
    if (user.token && user.type) {
      const url = proxy.wsNotification + "ws";
      try {
        socket.current = new WebSocket(url, user.token);
        // socket.current.onopen = (e) => console.log("onopen", e);
        socket.current.onmessage = (e) => {
          const newNotification = mapNotification(
            JSON.parse(e.data),
            user.type
          );
          setCount((prev) => prev + 1);
          setNotifications((prev) => [newNotification, ...prev]);
        };
        socket.current.onerror = (e) => {
          console.log(e);
          setError(true);
        };
      } catch (e) {
        console.log(e);
      }
    }
  }, [user.token, user.type]);

  const data = {
    notifications,
    count,
    error,
    next,
    socket,
    markAsRead: async () => {
      setCount(0);
      try {
        await markAsRead(user.token);
      } catch (e) {
        setError(true);
        return;
      }
    },
    deleteNotification: async (id) => {
      const notificationToRemove = notifications.find(
        (notification) => notification.id === id
      );
      if (notificationToRemove?.unread && count > 0) {
        setCount((prev) => prev - 1);
      }
      setNotifications((prev) =>
        prev.filter((notification) => notification.id !== id)
      );
      try {
        await deleteNotification(user.token, id);
      } catch (e) {
        setError(true);
        return;
      }
    },
    deleteNotifications: async () => {
      setCount(0);
      setNotifications([]);
      try {
        await deleteNotifications(user.token);
      } catch (e) {
        setError(true);
        return;
      }
    },
    loadMoreNotifications: async () => {
      setNext(null);
      let res;
      try {
        res = await getNotifications(user.token, user.type, next);
      } catch (e) {
        setError(true);
        return;
      }
      setNext(res.next);
      setNotifications((prev) => [...prev, ...res.notifications]);
    },
  };
  return <NotificationsContext.Provider value={data} {...props} />;
};
