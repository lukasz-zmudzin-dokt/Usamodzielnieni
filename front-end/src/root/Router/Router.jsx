import React from "react";
import { Route, Switch, useLocation } from "react-router-dom";
import routes from "constants/routes";
import PrivateRoute from "root/_components/PrivateRoute";
import { UserContext } from "context/UserContext";

const Router = () => {
  const location = useLocation();
  return (
    <UserContext.Consumer>
      {value => {
        return (
          <Switch location={location} key={location.pathname}>
            {routes.map(
              ({ component: Component, path, isPrivate, userVerified, ...rest }) => {
                if (isPrivate) {
                  return (
                    <PrivateRoute
                      path={path}
                      key={path}
                      userVerified={userVerified}
                      {...rest}
                      authenticated={value}
                    >
                      <Component {...rest} />
                    </PrivateRoute>
                  );
                }
                return (
                  <Route path={path} key={path} {...rest}>
                    <Component {...rest} />
                  </Route>
                );
              }
            )}
          </Switch>
        );
      }}
    </UserContext.Consumer>
  );
};

export default Router;
