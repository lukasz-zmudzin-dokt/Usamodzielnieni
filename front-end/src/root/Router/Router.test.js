import React from "react";
import { render } from "@testing-library/react";
import { createMemoryHistory } from "history";
import Router from "root/Router";
import { MemoryRouter, Router as RouterDOM } from "react-router-dom";
import { UserContext, AlertContext } from "context";

const renderWithRouter = (
  ui,
  context,
  contextA,
  {
    route = "/logina",
    history = createMemoryHistory({ initialEntries: [route] }),
  } = {}
) => {
  return {
    ...render(
      <UserContext.Provider value={context}>
        <AlertContext.Provider value={contextA}>
          <RouterDOM history={history}>{ui}</RouterDOM>
        </AlertContext.Provider>
      </UserContext.Provider>
    ),
    history,
  };
};

describe("Router test", () => {
  let context = {
    token: "abc",
  };

  let alertC = {
    showAlert: jest.fn(),
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("should match snapshot", () => {
    const { container } = render(
      <MemoryRouter>
        <Router></Router>
      </MemoryRouter>
    );
    expect(container).toMatchSnapshot();
  });
  it("full app rendering/navigating(dashboard)", () => {
    const history = createMemoryHistory();
    history.push("/");
    const { getByText } = render(
      <MemoryRouter>
        <Router></Router>
      </MemoryRouter>
    );

    expect(getByText("Utwórz konto")).toBeInTheDocument();
  });

  it("full app rendering/navigating(register)", () => {
    const { getByText } = render(
      <MemoryRouter initialEntries={["/newAccount"]}>
        <Router></Router>
      </MemoryRouter>
    );

    expect(getByText("Rejestracja")).toBeInTheDocument();
  });

  it("full app rendering/navigating(login)", () => {
    const { getByText } = render(
      <MemoryRouter initialEntries={["/login"]}>
        <Router></Router>
      </MemoryRouter>
    );

    expect(getByText("Logowanie")).toBeInTheDocument();
  });

  it("should redirect if route didn't exist(with token)", () => {
    renderWithRouter(<Router></Router>, context, alertC);

    expect(alertC.showAlert).toHaveBeenCalled();
    expect(alertC.showAlert).toHaveBeenCalledWith(
      "Nie ma takiej strony. Nastąpiło przekierowanie do twojego profilu."
    );
  });

  it("should redirect if route didn't exist(without token)", () => {
    context = {};
    renderWithRouter(<Router></Router>, context, alertC);

    expect(alertC.showAlert).toHaveBeenCalled();
    expect(alertC.showAlert).toHaveBeenCalledWith(
      "Nie ma takiej strony. Nastąpiło przekierowanie do strony głównej."
    );
  });
});
