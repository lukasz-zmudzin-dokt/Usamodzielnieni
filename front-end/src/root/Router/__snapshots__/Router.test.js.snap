// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Router test should match snapshot 1`] = `
<div>
  <div
    class="Menu container-fluid"
  >
    <div
      class="menu-background d-flex justify-content-center align-items-center"
    >
      <div
        class="row"
      >
        <div
          class="col"
        />
        <div
          class="col-7"
        >
          <img
            alt="Menu logo"
            class="menu-logo"
            src="logo.png"
          />
        </div>
        <div
          class="col"
        />
      </div>
      <div
        class="menu-button-row row"
      >
        <div
          class="col"
        />
        <div
          class="col"
        >
          <div
            class="btn-toolbar"
            role="toolbar"
          >
            <a
              class="menu-button-big btn btn-primary"
              href="/newAccount"
            >
              Utwórz konto
            </a>
            <a
              class="menu-button-big btn btn-primary"
              href="/login"
            >
              Zaloguj się
            </a>
          </div>
        </div>
        <div
          class="col"
        />
      </div>
      <div
        class="menu-button-row row"
      >
        <div
          class="col"
        />
        <div
          class="col"
        >
          <div
            class="btn-toolbar"
            role="toolbar"
          >
            <a
              class="menu-button-small menu-button-white btn btn-primary"
              href="/login"
            >
              Kreator CV
            </a>
            <a
              class="menu-button-small menu-button-white btn btn-primary"
              href="/myCVs"
            >
              Moje CV
            </a>
            <a
              class="menu-button-small menu-button-white btn btn-primary"
              href="/jobOffers"
            >
              Oferty pracy
            </a>
            <a
              class="menu-button-small menu-button-white btn btn-primary"
              href="/blog"
            >
              Blogi
            </a>
          </div>
        </div>
        <div
          class="col"
        />
      </div>
    </div>
  </div>
</div>
`;
